### ISSUE TITLE
Overzealous filename sanitation stripping valid starting and ending characters

### ISSUE THREAD

**Issue by Mosnar** [craftcms/cms#3608]
### Description
Files uploaded to Craft should preserve their original filenames as closely as possible. I'm not sure if there's a historical reason for it, but FileHelper.php is stripping out some perfectly valid characters from the start and end of filenames: `-_.`. In particular, we're running into pain where a client is trying to upload several hundred detailed product specifications using a format like: `123-4.5-__.pdf` to represent products in the 123-4.5 range. When an asset like this is uploaded, it is sanitized to `123-4.5.pdf` which is incorrect in this use case. I believe you could make a case for stripping out periods; however, I see no reason to force users to avoid dashes and underscores at the start and end of filenames.

I would propose at the very least, adding a configuration option to specify blacklisted starting and trailing characters.

### Additional info

- Craft version: 3.0.37
- PHP version: N/A
- Database driver & version: N/A
- Plugins & versions: N/A
- Host Platform: Amazon Linux/Mac OS

**Comment by Mosnar** [comment]
To save anyone else looking for this issue time, it looks like overriding this is probably impossible. My approach was to override the necessary controllers and helpers with subclasses. I first overrode **AssetsController.php** to catch the file upload via:
```php
'controllerMap' => [
        'assets' => \modules\mymodule\controllers\AssetsController::class,
    ],
```
and then I extended the assets helper and FileHelpers:
`class Assets extends \craft\helpers\Assets`
`class FileHelper extends \craft\helpers\FileHelper`

This gave me enough context to override the first instance of the sanitizeFilename method; however, the problem is that sanitizeFilename actually gets called again after the element is saved by the **AssetLocationValidator** referenced in the Asset element definition. At this point, the class is referenced explicitly and cannot be replaced.

**Comment by andris-sevcenko** [comment]
Looking at this code here: https://github.com/craftcms/cms/blob/develop/src/helpers/Assets.php#L135-L141

I think the best solution here is modifying the `craft\events\SetAssetFilenameEvent` class to include both the original filename and the sanitized version, firing that event after sanitizing the filename (as opposed to firing it before as it is currently) and allow plugins/modules to change the filename as they see fit.

That way all the sanitization still occurs, we're not adding new configuration options and we're providing the opportunity to modify the filename in the rare cases it's needed (seeing as how this is the first time it's been an issue).

It would have to happen in 3.1, though, as some plugins might be relying on that event to be fired before it's being sanitized, so it's a change some devs might need to look at.

Or add another event there that is fired after sanitization.

**Comment by andris-sevcenko** [comment]
Closed via #3615
